util/posix_close_normal.c

```diff
@@ -0,0 +1,129 @@
+#include <sys/socket.h>
+
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "noeintr.h"
+#include "warnp.h"
+
+#include "posix_close.h"
+
+/*
+ * The two files pthread_close_normal.c and pthread_close_pthread.c should be
+ * identical, other than the #define USE_PTHREAD line.
+ */
+#define USE_PTHREAD 0
+
+#if USE_PTHREAD
+#include <pthread.h>
+
+static pthread_mutex_t mutex;
+#endif
+
+#define MAGIC_COOKIE_LEN 4
+
+/**
+ * posix_close(fd, flags):
+ * Close a descriptor, but looping until completion if interrupted by a signal.
+ * ${flags} must be 0.
+ */
+int
+posix_close(int fd, int flags)
+{
+	uint8_t cookie_write[MAGIC_COOKIE_LEN];
+	uint8_t cookie_recv[MAGIC_COOKIE_LEN];
+	int s[2];
+	int i;
+	ssize_t r;
+
+	/* Sanity check. */
+	assert(flags == 0);
+
+	/* Create a socket pair. */
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, s)) {
```

Comment:
This needs to be done only once, and the socket pair is kept open for the lifetime of the process (or rather, until an atexit handler cleans it up).  Otherwise you have problems when you try to close the socket pair.

Reply:
Are you thinking of having an explicit `_init` function, or should this implicitly detect when it needs to be initialized?  (or both?)

Reply:
It should detect when it needs to be initialized.  Look at `pthread_once`.

Reply:
Hmm, that would presumably require linking to `pthread`.  I was going to do this instead:
```
static int s[2] = {-1, -1};

int
posix_close(int fd, int flags)
{
...
        /* Initialize (if needed). */
        if ((s[0] == -1) && posix_close_init()) {
                warn0("posix_close_init");
                goto err0;
        }
```

Reply:
And what happens if both threads run simultaneously and see `s[0] == -1`?

Doing initialization is *exactly why pthread_once exists*.

Reply:
But we're already wrapping the whole thing in a mutex.  And pthread_once() doesn't let us return a value [1].  This looks weird:

```
/* pthread_once() doesn't allow a return value from init_routine. */
static int init_ret = 0;

noeintr_close(int fd)
{
...
#if USE_PTHREAD
    if ((rc = pthread_mutex_lock(&mutex)) != 0) {
        ...
#endif

     /* Initiailize */
#if USE_PTHREAD
    if (pthread_once(&once_control, noeintr_close_init())) {
         ...
#else
    if (init_status == 0)
        noeintr_close_init();
#endif

    /* Check for any errors during initialization. */
    if (init_status != 1)
```

If we rely on the mutext to ensure that only one thread is running the code at once, then we can do:
```
noeintr_close(int fd)
{
...
#if USE_PTHREAD
    if ((rc = pthread_mutex_lock(&mutex)) != 0) {
        ...
#endif

     /* Initiailize */
    if ((init_status == 0) && (noeintr_close_init()) {
        warn0("noeintr_close_init");
        goto err1;
    }
```

which is much more idiomatic.  The init function has `socketpair`, `atexit`, `entropy_read`, and `noeintr_write`, so I don't think that checking its success is optional.

[1] https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_once.html